enhancements/insights/prometheus-alerts-for-insights-recommendations.md

```diff
@@ -0,0 +1,268 @@
+for additional information about recommendations to be implemented alongside
+what already exists.
+
+### API Extensions
+
+### Implementation Details/Notes/Constraints [optional]
+
+Please note that IO refers to the Insights Operator in this context.
+
+1. IO is enabled and periodically uploads archives to the Ingress service,
+   which are then processed by the CCX data pipeline. (standard behavior)
+2. After a successful archive upload, IO waits for the report to become
+   available, fetches it from the Smart Proxy and processes it. (This is an
+   extension of the current behavior, where we only look at the number of
+   active recommendations per severity level.) (needs to be implemented)
+3. In-cluster Prometheus metrics indicating the status of recommendations
+   being active are set accordingly.
+    - If a recommendation was previously registered as active and is present
+      in the latest Insights report, IO performs no additional action for this
+      recommendation.
+    - If a recommendation was previously registered as active and the latest
+      report no longer contains it, the metric will either be reset or
+      unregistered from the Prometheus instance if possible. (remove alerts for
+      no longer hitting recommendations)
+    - If a recommendation was previously registered as inactive or the metric
+      for this recommendation is absent and the latest Insights report contains
+      it, a metric will be updated/created to indicate that the recommendation
+      is currently hitting. (indicate new recommendations)
+4. The metric will have labels containing the ID, severity (based on the total
+   risk property), a human-readable name, and a link to a detailed description
+   of the recommendation and its remediation steps (Insights Advisor URL
+   generated based on the cluster ID and rule ID by IO) and its value will be
+   0/1 to indicate if the recommendation is inactive/active respectively.
+5. There will be a generic/templated info-level alert present on the cluster,
+   which will present the user with brief information about the hitting
+   recommendation (its human-readable name and severity),
+   and the link for details and remediation steps.
+
+It is worth pointing out that if any key part of the Prometheus stack is not
+running as expected on the cluster, then this feature will not work. Prometheus
+manages both the metrics and the alerts, so without it, no alerts can be
+generated by IO.
+
+**Unhealthy Insights Operator scenario:**
+
+- If something goes wrong with IO, we would not want the Insights alerts to
+  remain hanging on the cluster indefinitely.
+- The alert should only fire if IO has been healthy at any point in the past 10
+  minutes. Healthy, in this context, means that there is no major error
+  happening (crashlooping, degraded operator status, etc.). A temporary error
+  state should not cause the alerts to be disabled.
```

Comment:
You can get the fail-silent behavior by not serving the metrics unless you are happy with them.  You can bridge expected outages (e.g. insights pod restarts) with some over-time aggregation in the rule's `expr`.  I don't think you need anything fancier than that here; Prometheus and the Alertmanager will clear the firing alert if the metric that fed it goes away beyond the short interpolation window.

Reply:
This point was regarding an unhealthy IO state. If that happens, how would we remove the metrics? If IO is crashlooping, then it cannot really do anything about its metrics, right? So that's why I wanted an IO state check to be a part of the alerting expression.

Reply:
you don't have to remove metrics.  Unless you are actively serving them, [Prom will notice][1].

[1]: https://github.com/openshift/enhancements/pull/1036#discussion_r819175113